File: 0018-4sum/0018-4sum.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
            ans = set()
            if len(nums) < 4:
                return ans

            nums.sort()
            for i in range(len(nums)):
                for j in range(i+1, len(nums)):
                    l, r = j+1, len(nums) - 1
                    while l < r:
                        sm = nums[i]+nums[j]+nums[l]+nums[r]
                        if sm == target:
                            ans.add((nums[i], nums[j], nums[l], nums[r]))
                            l += 1

                        elif sm > target:
                            r -= 1
                        else:
                            l += 1

            return ans

File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        
        
	int n = (int)nums.size();
	if(n < 4)
		return {};

	sort(nums.begin(), nums.end());
	set<vector<int>> ans;
	long long x, sum;
	int i;
	for(i = 0 ; i < (int)nums.size(); i ++){
		int j;
		for(j = i+1; j < (int)nums.size(); j++){
			x = nums[i] + nums[j];

			int l = j + 1, r = n-1;
			while(l < r){
				sum = nums[l]+nums[r];

				if(sum + x == target){
					ans.insert({nums[i], nums[j], nums[l], nums[r]});
					l++;
				}else if(sum + x > target)
					r --;
				else
					l ++;
			}

		}
	}

	vector<vector<int>> res(ans.begin(), ans.end());
	return res;
        
        
        
    }
};